DuckDuckGo/Subscription/PrivacyProDataReporting.swift

```diff
@@ -0,0 +1,310 @@
+            case .appOnboardingCompleted: value = isAppOnboardingCompleted()
+            case .emailEnabled: value = isEmailEnabled()
+            case .widgetAdded: value = isWidgetAdded()
+            case .frequentUser: value = isFrequentUser()
+            case .longTermUser: value = isLongTermUser()
+            case .autofillUser: value = isAutofillUser()
+            case .validOpenTabsCount: value = isValidOpenTabsCount()
+            case .searchUser: value = isSearchUser()
+            }
+            additionalParameters[parameter.rawValue] = String(value)
+        }
+
+        return additionalParameters
+    }
+
+    func mergeRandomizedParameters(for useCase: UseCase, with parameters: [String: String]) -> [String: String] {
+        randomizedParameters(for: useCase).merging(parameters) { $1 }
+    }
+
+    func isReinstall() -> Bool {
+        _variantName == VariantIOS.returningUser.name
+    }
+
+    func isFireButtonUser() -> Bool {
+        _fireCount > Self.fireCountThreshold
+    }
+
+    func isSyncUsed() -> Bool {
+#if DEBUG
+        guard !ProcessInfo().arguments.contains("testing") else {
+            return false
+        }
+#endif
+        return _syncAuthState != .inactive
+    }
+
+    func isFireproofingUsed() -> Bool {
+        userDefaults.bool(forKey: Key.isFireproofingUsedKey, defaultValue: false)
+    }
+
+    func isAppOnboardingCompleted() -> Bool {
+        tutorialSettings.hasSeenOnboarding
+    }
+
+    func isEmailEnabled() -> Bool {
+        emailManager.isSignedIn
+    }
+
+    func isWidgetAdded() -> Bool {
+        userDefaults.bool(forKey: Key.isWidgetAddedKey, defaultValue: false)
+    }
+
+    func isFrequentUser() -> Bool {
+        let now = dateGenerator()
+        guard let _lastSessionEnded,
+              let daysSinceLastSession = Calendar.current.numberOfDaysBetween(_lastSessionEnded, and: now) else {
+            return false
+        }
+        return daysSinceLastSession < Self.frequentUserThreshold
+    }
+
+    func isLongTermUser() -> Bool {
+        guard let _installDate,
+              let daysSinceInstall = Calendar.current.numberOfDaysBetween(_installDate, and: dateGenerator()) else {
+            return false
+        }
+        return daysSinceInstall > Self.longTermUserThreshold
+    }
+
+    func isAutofillUser() -> Bool {
+        _accountsCount > Self.autofillUserThreshold
+    }
+
+    func isValidOpenTabsCount() -> Bool {
+         _tabsCount > Self.openTabsCountThreshold
+    }
+
+    func isSearchUser() -> Bool {
+        _searchCount > Self.searchCountThreshold
+    }
+
+    func saveFireproofingUsed() {
+        guard !isFireproofingUsed() else { return }
+        userDefaults.set(true, forKey: Key.isFireproofingUsedKey)
+    }
+
+    func saveWidgetAdded() async {
+        let isInstalled = await appSettings.isWidgetInstalled()
+        if isInstalled != isWidgetAdded() {
+            userDefaults.set(isInstalled, forKey: Key.isWidgetAddedKey)
```

Comment:
This is cached so that `isWidgetAdded()` doesn't need to be `async` and that reduces the complexity when call sites are updated.